docs/functions.md

```diff
@@ -0,0 +1,209 @@
+
+    ```
+    0000000000000000000000000000000000000000-1
+    ```
+
+    Whereas in _cluster_ mode it could be:
+
+    ```
+    a007297351b007297351c007297351d007297351-1
+    ```
+
+## Registration
+The representation of an event-driven function is called a registration.
+
+Registrations are persisted in Redis' snapshots, a.k.a RDB files. This allows recovering both data and event handlers in the database in the event of failure.
+
+!!! abstract "Related commands"
+    The following RedisGears commands are related to registering functions:
+
+    * [`RG.PYEXECUTE`](commands.md#rgpyexecute)
+    * [`RG.DUMPREGISTRATIONS`](commands.md#rgdumpregistrations)
+    * [`RG.UNREGISTER`](commands.md#rgunregister)
+
+## Registration ID
+Every registration has a unique internal identifier. It is generated in the same manner as the [Execution ID](#execution-id).
+
+## Context Builder
+RedisGears functions in Python always begin with a context builder - the [`#!python class GearsBuilder`](runtime.md#gearsbuilder).
+
+!!! tip
+    `GB()` is an alias for `GearsBuilder()`.
+
+    It is intended to be used for brevity, increased productivity and the reduction of finger strain due to repetitive typing.
+
+**Python API**
+```python
+class GearsBuilder(reader='KeysReader', defaultArg='*', desc=None)
+```
+_Arguments_
+
+* _reader_: the function's [reader](readers.md)
+* _defaultArg_: An optional argument that the reader may need. These are usually a key's name, prefix, glob-like or a regular expression. Its use depends on the function's reader type and action.
+* _desc_: an optional description
+
+**Examples**
+```python
+# Here's the default context builder being run
+GearsBuilder().run()
+
+# You can also do this
+gb = GB()
+gb.register()
+```
+
+## Actions
+An action is a special type of operation. It is always the function's final step.
+
+### Run
+The **Run** action runs a function as a batch. The function is executed once and exits once the data is exhausted by its reader.
+
+Trying to run more than one function in the same execution will fail with an error.
+
+!!! example "Multiple executions error"
+    ```
+    127.0.0.1:30001> RG.PYEXECUTE "GB().run()\nGB().run()"
+    (error) [... 'spam.error: Can not run more then 1 executions in a single script']
+    ```
+
+!!! important "Execution is always asynchronous"
+    Batch functions are **always executed asynchronously** by the RedisGears engine. That means that they are run in a background thread rather than by the main process of the Redis server.
+
+**Python API**
+```python
+class GearsBuilder.run(arg=None, convertToStr=True, collect=True)
+```
+
+_Arguments_
+
+* _arg_: An optional argument that's passed to the reader as its _defaultArg_. It means the following:
+    * A glob-like pattern for the [KeysReader](readers.md#keysreader) and [KeysOnlyReader](readers.md#keysonlyreaders) readers
+    * A key name for the [StreamReader](readers.md#streamreader) reader
```

Comment:
A python generator for PythonReader, Ignored by ShardsIdReader

Reply:
Fixed via 079f280